### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by raphaelrobert**

### CODE REVIEW

File: evercrypt-rs/benches/benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
+#[macro_use]
+extern crate criterion;
+extern crate evercrypt;
+extern crate rand;
+
+use criterion::Criterion;
+
+const DATA: &[u8; 1024] = &[1u8; 1024];
+
+pub fn randombytes(n: usize) -> Vec<u8> {
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:10]: I'll put this in the library later, using the Evercrypt randomness. But let's do it this way for now.

File: evercrypt-rs/benches/benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    use rand::rngs::OsRng;
+    use rand::RngCore;
+
+    let mut bytes = Vec::with_capacity(n);
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:14]: `let mut bytes = vec![0u8; n]` would be the nicer version here.

File: evercrypt-rs/benches/benchmark.rs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        let nonce = &randombytes(12);
+        let data = &randombytes(1_000);
+        let aad = &randombytes(1_000);
+        let aead = Aead::new(Mode::Aes128Gcm, key).unwrap();
+        b.iter(|| {
+            let (_ct, _tag) = aead.encrypt(data, &nonce, aad).unwrap();
+        });
+    });
+    c.bench_function("AES128 decrypt", |b| {
+        b.iter_with_setup(
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:76]: That is replaced with `iter_batched` in 0.3.

File: evercrypt-rs/benches/benchmark.rs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                let _decrypted = aead.decrypt(&ct, &tag, &nonce, &aad).unwrap();
+            },
+        )
+    });
+}
+
+fn criterion_x25519(c: &mut Criterion) {
+    use evercrypt::x25519::{x25519, x25519_base};
+    c.bench_function("X25519 base", |b| {
+        let sk = &randombytes(32);
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:153]: Should this go into a setup as well?

File: evercrypt-rs/benches/benchmark.rs [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+        b.iter(|| {
+            let _pk = x25519_base(sk);
+        });
+    });
+    c.bench_function("X25519 DH", |b| {
+        let sk1 = &randombytes(32);
+        let pk1 = x25519_base(sk1);
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:160]: I think using a setup and different keys here would be good as well.

File: evercrypt-rs/benches/benchmark.rs [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    c.bench_function("P256 DH", |b| {
+        let pk1 = hex_to_bytes(PK1_HEX);
+        let sk2 = hex_to_bytes(SK2_HEX);
+        b.iter(|| {
+            let _zz = p256_dh(&pk1, &sk2).unwrap();
+        });
+    });
+}
+
+fn criterion_ed25519(c: &mut Criterion) {
```
Comment (franziskuskiefer) [inline evercrypt-rs/benches/benchmark.rs:191]: Should these get setups as well?

**Comment by franziskuskiefer** [comment]
Oh, and could you add the benchmarks to the CI?

**Comment by raphaelrobert** [comment]
> Oh, and could you add the benchmarks to the CI?

Isn't that too heavy for the CI? The benchmarks take a while...

Also, where do I configure this?

**Comment by franziskuskiefer** [comment]
> > Oh, and could you add the benchmarks to the CI?
> 
> Isn't that too heavy for the CI? The benchmarks take a while...
> 
> Also, where do I configure this?

It's fine for GH actions I think. Just add another one here with `cargo bench` https://github.com/franziskuskiefer/evercrypt-rust/blob/master/.github/workflows/evercrypt-rs.yml#L39.
But I can also do that later if you prefer.

**Pull request merged**